Bottom: 7ad3499de95de793768f2232b79f7e5f84c4da3d
Top:    554a968418e246f9f3bb85aad8381468e856955a
Author: Jason Wang <jasowang@redhat.com>
Date:   2016-12-16 15:29:34 +0800

XDP: introduce XDP_HOLD


---

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_rx.c b/drivers/net/ethernet/mellanox/mlx4/en_rx.c
index 2264da0..c413871 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_rx.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_rx.c
@@ -926,6 +926,8 @@ int mlx4_en_process_rx_cq(struct net_device *dev, struct mlx4_en_cq *cq, int bud
 			switch (act) {
 			case XDP_PASS:
 				break;
+			case XDP_HOLD:
+				goto next;
 			case XDP_TX:
 				if (likely(!mlx4_en_xmit_frame(ring, frags, dev,
 							length, cq->ring,
diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 08327e0..5142e9c 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -374,6 +374,14 @@ static void virtnet_xdp_xmit(struct virtnet_info *vi,
 	virtqueue_kick(sq->vq);
 }
 
+static virtnet_xdp_buff_free(const void *r, const void *data)
+{
+	struct receive_queue *rq = r;
+	struct page *page = data;
+
+	give_pages(rq, page);
+}
+
 static u32 do_xdp_prog(struct virtnet_info *vi,
 		       struct receive_queue *rq,
 		       struct bpf_prog *xdp_prog,
@@ -394,6 +402,10 @@ static u32 do_xdp_prog(struct virtnet_info *vi,
 
 	xdp.data = buf + hdr_padded_len;
 	xdp.data_end = xdp.data + (len - vi->hdr_len);
+	xdp.ring = rq;
+	xdp.private = page;
+	xdp.free = virtnet_xdp_buff_free;
+	xdp.dev = vi->dev;
 
 	act = bpf_prog_run_xdp(xdp_prog, &xdp);
 	switch (act) {
@@ -406,6 +418,8 @@ static u32 do_xdp_prog(struct virtnet_info *vi,
 		xdp.data = buf + (vi->mergeable_rx_bufs ? 0 : 4);
 		virtnet_xdp_xmit(vi, rq, &vi->sq[qp], &xdp);
 		return XDP_TX;
+	case XDP_HOLD:
+		return XDP_HOLD;
 	default:
 		bpf_warn_invalid_xdp_action(act);
 	case XDP_ABORTED:
@@ -450,6 +464,9 @@ static struct sk_buff *receive_big(struct net_device *dev,
 			rcu_read_unlock();
 			goto xdp_xmit;
 		case XDP_DROP:
+		case XDP_HOLD:
+			rcu_read_unlock();
+			return NULL;
 		default:
 			goto err_xdp;
 		}
diff --git a/include/uapi/linux/bpf.h b/include/uapi/linux/bpf.h
index 0eb0e87..30d5238 100644
--- a/include/uapi/linux/bpf.h
+++ b/include/uapi/linux/bpf.h
@@ -595,6 +595,7 @@ enum xdp_action {
 	XDP_DROP,
 	XDP_PASS,
 	XDP_TX,
+	XDP_HOLD,
 };
 
 /* user accessible metadata for XDP packet hook
